src/cargo/core/resolver/mod.rs

```diff
@@ -585,35 +590,80 @@ fn activate(
     candidate: Candidate,
     method: &Method<'_>,
 ) -> ActivateResult<Option<(DepsFrame, Duration)>> {
+    let candidate_pid = candidate.summary.package_id();
     if let Some((parent, dep)) = parent {
-        cx.resolve_graph.push(GraphNode::Link(
-            parent.package_id(),
-            candidate.summary.package_id(),
-            dep.clone(),
-        ));
+        let parent_pid = parent.package_id();
+        cx.resolve_graph
+            .push(GraphNode::Link(parent_pid, candidate_pid, dep.clone()));
+        Rc::make_mut(
+            // add a edge from candidate to parent in the parents graph
+            cx.parents.link(candidate_pid, parent_pid),
+        )
+        // and associate dep with that edge
+        .push(dep.clone());
+        // For now the interaction of public dependency conflicts with renamed and
+        // 'cfg({})' dependencies is not clear. So we disable public checks if the
+        // other functionality is used.
+        // TODO: this disable is not a long term solution.
+        //       This needs to be removed before public dependencies are stabilized!
+        if dep.platform().is_none() && dep.explicit_name_in_toml().is_none() {
```

Comment:
I think the behavior that we want is to probably just delete this `if` clause and unconditionally run the below?

Reply:
Yes, we would like to delete this if statement and have the code below deal with the special handling of these corner cases. At the beginning of this PR we did not have this if statement and it broke the existing tests. Until we figure out how the corner cases are supposed to be handled, this if statement lets other people make progress.

Reply:
Er sorry it's taking me awhile to get back to this, but do you have a corner case in mind that this is needed for? I can't actually think of a case where the platform and/or renaming would cause issues

Reply:
It is easy to find the examples, just look at the test that do not pass without it.
- renaming: [this test](https://github.com/rust-lang/cargo/blob/d75d1fbe6b43e59d0a04e73659e411855c2633fc/tests/testsuite/rename_deps.rs#L7) means that `foo` can see `bar 0.1.0` and `bar 0.2.0`. `foo` can get `expected bar::Thing but got bar::Thing` errores so is not permitted by public & private dependencies.
- platform: [this test](https://github.com/rust-lang/cargo/blob/1e768f712dfac8c093d46910a7582a3aec0e03de/tests/testsuite/cfg.rs) means that `foo` can see `bar 0.1.0` and `bar 0.2.0` depending on platform. This should not be a problem, but requires a deep level of disjointness analysis about platforms to be added to make it work.

Reply:
For renaming though we want to give an error in that case, so renaming doesn't affect public/private, right? For platforms it's true that in theory we want to take the exhaustive list of any platforms and if none of them actually activate both dependencies we don't give an error, but it seems that for all practical purposes we'd want to give an error assuming all dependencies can be activated on at least one platform.

Reply:
That is a reasonable position. (although different from what you suggested at https://github.com/rust-lang/cargo/issues/6129#issuecomment-465189082) It would mean that this existing stable behavior should not be allowed. Such a decision would require more discussion. I thought that we could have that discussion in a follow up PR if that is the behavior we decide on. 

This PR is the minimum functionality to unlock further work. I thought it would be more in kind for this PR to adopt the behavior of treating `cfg` and `rename` as having the "escape hatch" you suggested.